cc/crypto/hpke/sender_context.cc

```diff
@@ -0,0 +1,197 @@
+  if (!EVP_HPKE_CTX_seal(
+          /* ctx= */ hpke_context_.get(),
+          /* out= */ encrypted_message.data(),
+          /* out_len= */ &encrypted_message_len,
+          /* max_out_len= */ max_out_len,
+          /* in= */ plaintext_bytes.data(),
+          /* in_len= */ plaintext_bytes.size(),
+          /* ad= */ associated_data_bytes.data(),
+          /* ad_len= */ associated_data_bytes.size())) {
+    return absl::AbortedError("Failed to seal request");
+  }
+  encrypted_message.resize(encrypted_message_len);
+
+  std::string serialized_encrypted_message(encrypted_message.begin(), encrypted_message.end());
+  return serialized_encrypted_message;
+}
+
+SenderRequestContext::~SenderRequestContext() { EVP_HPKE_CTX_free(hpke_context_.release()); }
+
+absl::StatusOr<std::string> SenderResponseContext::Open(absl::string_view ciphertext,
+                                                        absl::string_view associated_data) {
+  std::vector<uint8_t> ciphertext_bytes = StringViewToBytes(ciphertext);
+  if (ciphertext_bytes.size() == 0) {
+    return absl::InvalidArgumentError("No ciphertext was provided.");
+  }
+  std::vector<uint8_t> associated_data_bytes = StringViewToBytes(associated_data);
+
+  // The plaintext should not be longer than the ciphertext.
+  std::vector<uint8_t> plaintext_bytes(ciphertext_bytes.size());
+  size_t plaintext_bytes_size;
+
+  if (!EVP_AEAD_CTX_open(
+          /* ctx= */ aead_response_context_.get(),
+          /* out= */ plaintext_bytes.data(),
+          /* out_len= */ &plaintext_bytes_size,
+          /* max_out_len= */ ciphertext_bytes.size(),
+          /* nonce= */ response_nonce_.data(),
+          /* nonce_len= */ response_nonce_.size(),
+          /* in= */ ciphertext_bytes.data(),
+          /* in_len= */ ciphertext_bytes.size(),
+          /* ad= */ associated_data_bytes.data(),
+          /* ad_len= */ associated_data_bytes.size())) {
+    return absl::AbortedError("Unable to decrypt response message");
+  }
+  plaintext_bytes.resize(plaintext_bytes_size);
+  std::string plaintext(plaintext_bytes.begin(), plaintext_bytes.end());
+  return plaintext;
+}
+
+SenderResponseContext::~SenderResponseContext() {
+  EVP_AEAD_CTX_free(aead_response_context_.release());
+}
+
+absl::StatusOr<SenderHPKEInfo> SetUpBaseSender(absl::string_view serialized_recipient_public_key,
+                                               absl::string_view info) {
+  SenderHPKEInfo sender_hpke_info;
+
+  // First collect encapsulated public key information and sender request context.
+  KeyInfo encap_public_key_info;
+  encap_public_key_info.key_bytes = std::vector<uint8_t>(EVP_HPKE_MAX_ENC_LENGTH);
+
+  std::vector<uint8_t> recipient_public_key_bytes =
+      StringViewToBytes(serialized_recipient_public_key);
+
+  if (recipient_public_key_bytes.size() == 0) {
+    return absl::InvalidArgumentError("No key was provided.");
+  }
+
+  std::vector<uint8_t> info_bytes = StringViewToBytes(info);
```

Comment:
Can we just get a pointer out of the `absl::string_view` ?

Reply:
Yes! I didn't realize we could before. I made that change here and elsewhere in the file.